uplink/src/base/bridge/actions_lane.rs

```diff
@@ -0,0 +1,915 @@
+    NoRoute(String),
+    #[error("Action timedout")]
+    ActionTimeout,
+    #[error("Another action is currently being processed")]
+    Busy,
+}
+
+struct RedirectionError(Action);
+
+pub struct ActionsBridge {
+    /// All configuration
+    config: Arc<Config>,
+    /// Tx handle to give to apps
+    status_tx: Sender<ActionResponse>,
+    /// Rx to receive action status from apps
+    status_rx: Receiver<ActionResponse>,
+    /// Actions incoming from backend
+    actions_rx: Receiver<Action>,
+    /// Contains stream to send ActionResponses on
+    streams: Streams,
+    /// Apps registered with the bridge
+    /// NOTE: Sometimes action_routes could overlap, the latest route
+    /// to be registered will be used in such a circumstance.
+    action_routes: HashMap<String, ActionRouter>,
+    /// Action redirections
+    action_redirections: HashMap<String, String>,
+    /// Current action that is being processed
+    current_action: Option<CurrentAction>,
+    parallel_actions: HashSet<String>,
+    ctrl_rx: Receiver<ActionBridgeShutdown>,
+    ctrl_tx: Sender<ActionBridgeShutdown>,
+    shutdown_handle: Sender<()>,
+}
+
+impl ActionsBridge {
+    pub fn new(
+        config: Arc<Config>,
+        package_tx: Sender<Box<dyn Package>>,
+        actions_rx: Receiver<Action>,
+        shutdown_handle: Sender<()>,
+        metrics_tx: Sender<StreamMetrics>,
+    ) -> Self {
+        let (status_tx, status_rx) = bounded(10);
+        let action_redirections = config.action_redirections.clone();
+        let (ctrl_tx, ctrl_rx) = bounded(1);
+
+        let mut streams_config = HashMap::new();
+        streams_config.insert("action_status".to_owned(), config.action_status.clone());
+        let mut streams = Streams::new(config.clone(), package_tx, metrics_tx);
+        streams.config_streams(streams_config);
+
+        Self {
+            status_tx,
+            status_rx,
+            config,
+            actions_rx,
+            streams,
+            action_routes: HashMap::with_capacity(10),
+            action_redirections,
+            current_action: None,
+            parallel_actions: HashSet::new(),
+            shutdown_handle,
+            ctrl_rx,
+            ctrl_tx,
+        }
+    }
+
+    pub fn register_action_route(&mut self, route: ActionRoute) -> Receiver<Action> {
+        let (actions_tx, actions_rx) = bounded(1);
+        self.insert_route(route, actions_tx.clone());
+
+        actions_rx
+    }
+
+    pub fn register_action_routes<R: Into<ActionRoute>, V: IntoIterator<Item = R>>(
+        &mut self,
+        routes: V,
+    ) -> Option<Receiver<Action>> {
+        let routes: Vec<ActionRoute> = routes.into_iter().map(|n| n.into()).collect();
+        if routes.is_empty() {
+            return None;
+        }
+
+        let (actions_tx, actions_rx) = bounded(1);
+        for route in routes {
+            self.insert_route(route, actions_tx.clone());
+        }
+
+        Some(actions_rx)
+    }
+
+    fn insert_route(
```

Comment:
This is only being used in the above 2 methods. Move this implementation to `register_action_route` and use `register_action_route` in `register_action_routes`